base/utfcheck.jl

```diff
@@ -0,0 +1,189 @@
+                    byt, pos = next(dat, pos)
+                    (byt != 0xed) && throw(UnicodeError(UTF_ERR_NOT_TRAIL, pos, byt))
+                    byt, pos = next(dat, pos)
+                    surr = get_continuation(0x0000d, byt, pos)
+                    byt, pos = next(dat, pos)
+                    surr = get_continuation(surr, byt, pos)
+                    !is_surrogate_trail(surr) && throw(UnicodeError(UTF_ERR_NOT_TRAIL, pos-2, surr))
+                    !accept_surrogates && throw(UnicodeError(UTF_ERR_SURROGATE, pos-2, surr))
+                    flags |= UTF_SURROGATE
+                    num4byte += 1
+                elseif ch > 0x07ff
+                    num3byte += 1
+                elseif accept_long_char
+                    flags |= UTF_LONG
+                    num2byte += 1
+                else
+                    throw(UnicodeError(UTF_ERR_LONG, pos-2, ch))
+                end
+            elseif ch < 0xf5
+                # 4-byte UTF-8 sequence (i.e. characters > 0xffff)
+                (pos + 2 > endpos) && throw(UnicodeError(UTF_ERR_SHORT, pos, ch))
+                byt, pos = next(dat, pos)
+                ch = get_continuation(ch & 0x07, byt, pos)
+                byt, pos = next(dat, pos)
+                ch = get_continuation(ch, byt, pos)
+                byt, pos = next(dat, pos)
+                ch = get_continuation(ch, byt, pos)
+                if ch > 0x10ffff
+                    throw(UnicodeError(UTF_ERR_INVALID, pos-3, ch))
+                elseif ch > 0xffff
+                    num4byte += 1
+                elseif is_surrogate_codeunit(ch)
+                    throw(UnicodeError(UTF_ERR_SURROGATE, pos-3, ch))
+                elseif accept_long_char
+                    # This is an overly long encoded character
+                    flags |= UTF_LONG
+                    if ch > 0x7ff
+                        num3byte += 1
+                    elseif ch > 0x7f
+                        num2byte += 1
+                    end
+                else
+                    throw(UnicodeError(UTF_ERR_LONG, pos-2, ch))
+                end
+            else
+                throw(UnicodeError(UTF_ERR_INVALID, pos, ch))
+            end
+        end
+    end
+    num3byte != 0 && (flags |= UTF_UNICODE3)
+    num4byte != 0 && (flags |= UTF_UNICODE4)
+    return totalchar, flags, num4byte, num3byte, num2byte
+end
+
+function check_string{T <: Union(Vector{UInt16}, Vector{UInt32}, AbstractString)}(
+                      dat::T,
+                      endpos = endof(dat),
+                      pos = start(dat)
+                      ;
+                      accept_long_null  = true,
+                      accept_surrogates = true,
+                      accept_long_char  = false)
+    local ch::UInt32
+    flags::UInt = 0
+    totalchar = num2byte = num3byte = num4byte = 0
+    @inbounds while pos <= endpos
```

Comment:
Since `pos` and `endpos` are caller-provided, you need to check `pos >= start(dat)` and `endpos <= endof(dat)` before the loop body and throw a `BoundsError()`, in order to avoid crashing on invalid arguments due to `@inbounds`.